### PR TITLE
Add Digest AKA to Algorithm Directive from RFC 3110 and RFC 4169

### PR DESCRIPTION
**Pull request by tobiasfunke1**
Hi,

I added changes to match the definition for the algorithm directive from RFC 3310:

```
           algorithm           =  "algorithm" EQUAL ( aka-namespace
                                  / algorithm-value )
           aka-namespace       =  aka-version "-" algorithm-value
           aka-version         =  "AKAv" 1*DIGIT
           algorithm-value     =  ( "MD5" / "MD5-sess" / token )
```

~~Since the no Digest AKA is supported right now, I decided to return an error on the `handle_digest_challenge` method.~~ I added a second commit to add Digest AKA in the `handle_digest_challenge` method too. 

Additional I decided to add some tests to check if the parsing works fine.

AKAv1: https://datatracker.ietf.org/doc/html/rfc3310#section-3.1
AKAv2: https://datatracker.ietf.org/doc/html/rfc4169#section-2

### CODE REVIEW

File: crates/sip-auth/src/digest.rs [hunk 1]
```diff
@@ -162,26 +162,33 @@ impl DigestAuthenticator {
         digest: DigestChallenge,
         request_parts: RequestParts<'_>,
     ) -> Result<AuthResponse, Error> {
-        let (hash, is_session): (HashFn, bool) = match digest.algorithm {
-            Algorithm::MD5 => {
+        // Get the algorithm
```
Comment (tobiasfunke1) [inline crates/sip-auth/src/digest.rs:165]:
```suggestion
```

File: crates/sip-auth/src/digest.rs [hunk 2]
```diff
@@ -162,26 +162,33 @@ impl DigestAuthenticator {
...
+        let algorithm_value = match digest.algorithm.clone() {
+            Algorithm::AkaNamespace((_, av)) => av,
+            Algorithm::AlgorithmValue(av) => av,
+        };
+
+        // Select the hash function based on the algorithm
```
Comment (tobiasfunke1) [inline crates/sip-auth/src/digest.rs:171]:
```suggestion
```

File: crates/sip-types/src/header/typed/auth.rs [hunk 3]
```diff
@@ -457,6 +457,72 @@ impl Display for QopOption {
...
 pub enum Algorithm {
+    AkaNamespace((AkaVersion, AlgorithmValue)),
+    AlgorithmValue(AlgorithmValue),
+}
+
+impl From<BytesStr> for Algorithm {
+    fn from(value: BytesStr) -> Self {
+        // Check if it is an aka-namespace
+        if let Some((version, algorithm)) = value.split_once('-') {
+            // Needs to have 5 characters and start with case-insensitive "AKAv"
```
Comment (tobiasfunke1) [inline crates/sip-types/src/header/typed/auth.rs:468]:
```suggestion
            // 5 characters, starting with case-insensitive "AKAv", followed by single-digit number
```

**Review by tobiasfunke1 — COMMENTED** [review]
Update and remove the comments

**Comment by tobiasfunke1** [comment]
I clean up the code and the commits, this PR is ready to review. @kbalt If you have time, can you please look at it?

For Digest AKA the password used to generate the digest response can contain non-ascii chars. So I change the type of the password to a vector of u8.
```
...
The resulting AKA RES parameter is treated as a "password" when
   calculating the response directive of [RFC 2617](https://datatracker.ietf.org/doc/html/rfc2617).
```
source: https://datatracker.ietf.org/doc/html/rfc3310#section-3.3

**Comment by tobiasfunke1** [comment]
Sorry, I accidentally closed this PR with a force push. Now it should be open again.

**Comment by kbalt** [comment]
Thanks for your effort, I will look into this when I have a bit of time to check everything out!

**Comment by tobiasfunke1** [comment]
Any updates on the review process?

**Pull request closed without merging**